src/Compilers/Core/Portable/CommandLine/EditorConfig.cs

```diff
@@ -0,0 +1,167 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.IO;
+using System.Text.RegularExpressions;
+
+namespace Microsoft.CodeAnalysis
+{
+    /// <summary>
+    /// Represents a single EditorConfig file, see http://editorconfig.org for details about the format.
+    /// </summary>
+    internal sealed class EditorConfig
+    {
+        // Matches EditorConfig section header such as "[*.{js,py}]", see http://editorconfig.org for details
+        private static readonly Regex _sectionMatcher = new Regex(@"^\s*\[(([^#;]|\\#|\\;)+)\]\s*([#;].*)?$", RegexOptions.Compiled);
+        // Matches EditorConfig property such as "indent_style = space", see http://editorconfig.org for details
+        private static readonly Regex _propertyMatcher = new Regex(@"^\s*([\w\.\-_]+)\s*[=:]\s*(.*?)\s*([#;].*)?$", RegexOptions.Compiled);
+
+        public Section GlobalSection { get; }
+
+        public string Directory { get; }
+
+        public ImmutableArray<Section> NamedSections { get; }
+
+        private EditorConfig(Section globalSection, ImmutableArray<Section> namedSections, string directory)
+        {
+            GlobalSection = globalSection;
+            NamedSections = namedSections;
+            Directory = directory;
+        }
+
+        /// <summary>
+        /// Gets whether this editorconfig is a the topmost editorconfig.
+        /// </summary>
+        public bool IsRoot
+            => GlobalSection.Properties.TryGetValue("root", out string val) && val == "true";
+
+        /// <summary>
+        /// Parses an editor config file text located within the given parent directory.
+        /// </summary>
+        public static EditorConfig Parse(string text, string parentDirectory)
+        {
+            Section globalSection = null;
+            var namedSectionBuilder = ImmutableArray.CreateBuilder<Section>();
+
+            // N.B. The editorconfig documentation is quite loose on property interpretation.
+            // Specifically, it says:
+            //      Currently all properties and values are case-insensitive.
+            //      They are lowercased when parsed.
+            // To accomodate this, we use a lower case Unicode mapping when adding to the
+            // dictionary, but we also use a case-insensitve key comparer when doing lookups
```

Comment:
❓ What will happen to this?

https://github.com/pkzxs/Aurora.Music/blob/8b6a97f9d8457b834dc86eea573c75410ea8dea7/Source/.editorconfig#L18

Reply:
It will be lowercased, as required by the spec.

Reply:
That will break existing roslyn IDE scenarios that customers are already depending on... Can we not lowercase and allow clients of the parser to decide the semantics they want for this sort of thing?

Reply:
If we end up using the case sensitive version there was no point in lowercasing it anyway. If we want to be non-compliant we should do so strongly and clearly.

Reply:
> If we want to be non-compliant we should do so strongly and clearly.

That was my intent when i had CaseSensitiveProperties and CaseInsensitiveProperties.  It makes it very clear what you're getting, and anyone consuming this has to consider which is the right thing for their domain.

Reply:
Note: i can see info about things being case insensitive here: https://github.com/editorconfig/editorconfig/wiki/EditorConfig-Properties

But it makes it sound more like normative text letting you know that for those specific properties, things are case insensitive, not necessarily a blanket rule that all property values are case insensitive.

Reply:
@sharwell So the parser should recognize the properties listed on the website and lowercase them, but everything else is our decision? And any new properties get added to the special list?

@CyrusNajmabadi You also intend this to apply to keys? Or just values?

Reply:
This link (http://editorconfig.org/#file-format-details) happens to mention: ```Currently all properties and values are case-insensitive. They are lowercased when parsed.```

Reply:
@agocke That's what this library does:

https://github.com/editorconfig/editorconfig-core-net/blob/master/src/EditorConfig.Core/FileConfiguration.cs#L95

lowercases the specified ones, does not touch the rest.  That's the same with this js impl: https://github.com/editorconfig/editorconfig-core-js/blob/master/src/index.ts#L125

Reply:
Note: thse were two actual impls linked to from the main editorconfig page.  So i thought i would take a look to see what the rest of the ecosystem is doing.

Reply:
@CyrusNajmabadi Also interesting to note that keys are always converted to lowercase. Only the values get special handling depending on the key name.

Reply:
C version also only lowercases the values of the well known set: https://github.com/editorconfig/editorconfig-core-c/blob/master/src/lib/editorconfig.c#L84

Reply:
Given that this will ship in the compiler, what kind of update contract should apply?

Reply:
> Given that this will ship in the compiler, what kind of update contract should apply?

Proposal:

* Keys are case insensitive and converted to lowercase when parsed
* Values are generally case sensitive, and trimmed but not otherwise modified during parsing
* Values which are defined by editorconfig.org are converted to lowercase if specified by that site, except for cases where the key contains a `.` (no such key currently falls into this exception, but it allows for changes to be made if the core specification changes while still allowing users to define stable property behavior via "namespacing")

Reply:
That's not an update contract. What happens when the website is updated? Do we promise to update our list? In what timeframe?

Reply:
It woudl be interesting if they updated in a substantial manner.  My personal preference would be that if such a change would cause a break in our existing behavior (i.e. existing keys/values would be interpretted differently), then we would have ways to ask for the specific formats.  

i.e. you could ask for the V1 EditorConfig interpretation (which we would always preserve), or you could ask for the V2 if that comes about.  etc. etc.

Reply:
I've implemented a system of "reserved" keys and values similar to the other implementations. Right now I provide no update contract, only saying that things may be added as we go on. Also, there is no special handling of "." as I didn't think that additional level of complexity was merited.